### PR TITLE
Now you can use vcpkg to get some of the dependencies.

### PR DESCRIPTION
**Pull request by TerensTare**
This feature can be enabled by defining the `USE_VCPKG` flag when running `cmake ..` (so instead of just `cmake ..`, do `cmake -DUSE_VCPKG=ON ..`). Tested on Windows 10 with Visual Studio 2019 and the latest vcpkg commit.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,30 +1,53 @@
...
-add_subdirectory(extlibs/lua)
-add_subdirectory(extlibs/sol3)
-
 # Audio library
 add_subdirectory(extlibs/soloud)
 
-if (APPLE OR USE_FILESYSTEM_FALLBACK)
+if (APPLE OR USE_FILESYSTEM_FALLBACK AND NOT OBE_USE_VCPKG)
     add_subdirectory(extlibs/tinydir)
     set(USE_FILESYSTEM_FALLBACK TRUE CACHE BOOL "Use filesystem fallback" FORCE)
```
Comment (Sygmei) [inline CMakeLists.txt:52]: So if we use VCPKG the flag won't be set ?

File: CMakeLists.txt [hunk 2]
```diff
@@ -33,9 +56,43 @@ endif()
...
+
+
+    link_libraries(${LUA_LIBRARIES})
+    include_directories(${LUA_INCLUDE_DIR})
+
+    link_libraries(sol2::sol2)
+    link_libraries(fmt::fmt fmt::fmt-header-only)
+    link_libraries(spdlog::spdlog spdlog::spdlog_header_only)
+    link_libraries(Catch2::Catch2)
+    else() # if using built-in dependencies
```
Comment (Sygmei) [inline CMakeLists.txt:89]: Wrong indentation ?

File: CMakeLists.txt [hunk 3]
```diff
@@ -33,9 +56,43 @@ endif()
...
+        link_libraries(FLAC OpenAL OpenGL Vorbis)
+    else()
+        link_libraries(X11 FLAC UDev OpenAL)
+    endif()
+
+
+    link_libraries(${LUA_LIBRARIES})
+    include_directories(${LUA_INCLUDE_DIR})
+
+    link_libraries(sol2::sol2)
```
Comment (Sygmei) [inline CMakeLists.txt:85]: Can we use `target_link_libraries` instead ?

File: CMakeLists.txt (hunk 1 above)
Comment (TerensTare) [reply CMakeLists.txt:52]: I added that here because Vcpkg has support for tinydir

File: CMakeLists.txt (hunk 2 above)
Comment (TerensTare) [reply CMakeLists.txt:89]: Yes, my bad sorry.

File: CMakeLists.txt (hunk 3 above)
Comment (TerensTare) [reply CMakeLists.txt:85]: That is what I initially used, but I got an error saying that the target wasn't built by this CMakeLists.txt

File: CMakeLists.txt (hunk 1 above)
Comment (Sygmei) [reply CMakeLists.txt:52]: If `OBE_USE_VCPKG` is set to `TRUE` the line
```cmake
set(USE_FILESYSTEM_FALLBACK TRUE CACHE BOOL "Use filesystem fallback" FORCE)
```
will never be executed.
Then, in [src/Core/CMakeLists.txt#L38](https://github.com/Sygmei/ObEngine/blob/master/src/Core/CMakeLists.txt#L38) it will fail to do the appropriate actions.

Comment (TerensTare) [reply CMakeLists.txt:52]: Fix coming right now

File: CMakeLists.txt (hunk 3 above)
Comment (Sygmei) [reply CMakeLists.txt:85]: Yup, the `target_link_libraries` are done in a separate [`CMakeLists.txt`](https://github.com/Sygmei/ObEngine/blob/master/src/Core/CMakeLists.txt#L26) (which is the CMakeLists.txt for ÖbEngineCore).

Comment (TerensTare) [reply CMakeLists.txt:86]: Hmm.. How can we make the other `CMakeLists.txt` know that we are using vcpkg?

Comment (Sygmei) [reply CMakeLists.txt:86]: Variables are shared between the `CMakeLists.txt` so `OBE_USE_VCPKG` should be accessible from there

File: src/Core/CMakeLists.txt [hunk 4]
```diff
@@ -23,21 +23,47 @@ target_include_directories(ObEngineCore
...
+    else()
+        target_link_libraries(ObEngineCore X11 FLAC UDev OpenAL)
+    endif()
+
+
+    target_link_libraries(ObEngineCore ${LUA_LIBRARIES})
+    target_include_directories(ObEngineCore ${LUA_INCLUDE_DIR})
+
+    target_link_libraries(ObEngineCore sol2::sol2)
+    target_link_libraries(ObEngineCore fmt::fmt fmt::fmt-header-only)
```
Comment (Sygmei) [inline src/Core/CMakeLists.txt:42]: Is there a non-header-only version available ?

Comment (TerensTare) [reply src/Core/CMakeLists.txt:42]: I think vcpkg supports only the header-only version.

Comment (TerensTare) [reply src/Core/CMakeLists.txt:42]: At least, on my pc

Comment (Sygmei) [reply src/Core/CMakeLists.txt:42]: It seems you are linking the two versions of the library, the standard one and the header only version.
Check out this link : https://fmt.dev/dev/usage.html#usage-with-cmake
Does removing the `fmt::fmt-header-only` and keeping `fmt::fmt` works ?

Comment (TerensTare) [reply src/Core/CMakeLists.txt:42]: Oh, I thought you were talking about sol2. I guess fmt will still work, there are places on my project where I use `fmt/core.h`, or `fmt/format.h` alone, so there are multiple headers.

Comment (Sygmei) [reply src/Core/CMakeLists.txt:42]: Same for `spdlog` :) I'd rather use the non-header-only version for faster build times

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,41 +1,99 @@
 cmake_minimum_required(VERSION 3.16)
 set(OpenGL_GL_PREFERENCE GLVND)
 
+option(USE_VCPKG "Use vcpkg to handle some of the dependencies" OFF)
```
Comment (Sygmei) [inline CMakeLists.txt:4]: Would it be possible to wrap the whole VCPKG initialization into a script in the [`cmake`](https://github.com/Sygmei/ObEngine/tree/master/cmake) directory ?

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,41 +1,99 @@
...
+    find_package(lua REQUIRED)
+    find_package(sol2 CONFIG REQUIRED)
+    find_package(fmt CONFIG REQUIRED)
+    find_package(spdlog CONFIG REQUIRED)
+    find_package(Catch2 CONFIG REQUIRED)
+    find_path(TINYDIR_INCLUDE_DIRS "tinydir.h")
+
+    find_package(SFML COMPONENTS system window graphics CONFIG REQUIRED)
+
+    include_directories(${TINYDIR_INCLUDE_DIRS})
```
Comment (Sygmei) [inline CMakeLists.txt:74]: Can we remove those now that they have been added into the [`src/Core/CMakeLists.txt`](https://github.com/Sygmei/ObEngine/blob/master/src/Core/CMakeLists.txt) ?

File: CMakeLists.txt [hunk 7]
```diff
@@ -40,40 +43,19 @@ endif()
...
-    find_package(fmt CONFIG REQUIRED)
-    find_package(spdlog CONFIG REQUIRED)
-    find_package(Catch2 CONFIG REQUIRED)
-    find_path(TINYDIR_INCLUDE_DIRS "tinydir.h")
-
-
-    include_directories(${TINYDIR_INCLUDE_DIRS})
-
     if(WIN32)
         link_libraries(FLAC OpenAL OpenGL Vorbis)
```
Comment (Sygmei) [inline CMakeLists.txt:52]: `target_link_libraries` :) ?

File: extlibs/vili/CMakeLists.txt [hunk 8]
```diff
@@ -28,7 +28,14 @@ set(VILI_SOURCES
 add_library(vili ${VILI_HEADERS} ${VILI_SOURCES})
 
 target_link_libraries(vili fifo_map)
-target_link_libraries(vili fmt)
+
+if(OBE_USE_VCPKG)
+    find_package(fmt CONFIG REQUIRED)
```
Comment (Sygmei) [inline extlibs/vili/CMakeLists.txt:33]: Do you really need to do a `find_package` again ?

File: src/Core/CMakeLists.txt [hunk 9]
```diff
@@ -24,40 +24,33 @@ target_include_directories(ObEngineCore
 )
 
 if(OBE_USE_VCPKG)
+    find_package(fmt CONFIG REQUIRED)
```
Comment (Sygmei) [inline src/Core/CMakeLists.txt:27]: `find_package` in top-level CMakeLists.txt and then use the with `target_link_libraries` on targets ?

File: src/Core/CMakeLists.txt [hunk 10]
```diff
@@ -24,40 +24,33 @@ target_include_directories(ObEngineCore
...
+    find_package(spdlog CONFIG REQUIRED)
+
+    find_package(sol2 CONFIG REQUIRED)
+    find_package(Catch2 CONFIG REQUIRED)
+
+    target_include_directories(ObEngineCore PUBLIC
```
Comment (Sygmei) [inline src/Core/CMakeLists.txt:33]: Why is this needed ? `target_link_libraries` should add the necessary include paths to the target ?

File: CMakeLists.txt (hunk 7 above)
Comment (TerensTare) [reply CMakeLists.txt:52]: It's on the main file, but I doubt it will need those checks and links.

File: extlibs/vili/CMakeLists.txt (hunk 8 above)
Comment (TerensTare) [reply extlibs/vili/CMakeLists.txt:33]: Yes, you need it there unfortunately. A wave of errors followed when that wasn't set. I can try to remove that from the Core CmakeList

File: CMakeLists.txt [hunk 11]
```diff
@@ -1,40 +1,50 @@
...
+
+    # fmtlib is required by spdlog
+    add_subdirectory(extlibs/fmt)
+    add_subdirectory(extlibs/spdlog)
+
+    # lua is required by sol
+    add_subdirectory(extlibs/lua)
+    add_subdirectory(extlibs/sol3)
+    
+    add_subdirectory(extlibs/tinydir)
```
Comment (Sygmei) [inline CMakeLists.txt:27]: We should check the condition
```cmake
if (APPLE OR USE_FILESYSTEM_FALLBACK)
```
there too

File: CMakeLists.txt [hunk 12]
```diff
@@ -1,40 +1,50 @@
...
     set(USE_FILESYSTEM_FALLBACK TRUE CACHE BOOL "Use filesystem fallback" FORCE)
 endif()
 
 # minizip is required by elzip
 add_subdirectory(extlibs/minizip)
 add_subdirectory(extlibs/elzip)
 
-# SFML 2.5 is minimum version
-find_package(SFML 2.5 COMPONENTS system window graphics network audio REQUIRED)
+find_package(SFML COMPONENTS system window graphics CONFIG REQUIRED)
```
Comment (Sygmei) [inline CMakeLists.txt:46]: Why don't we require the version 2.5 anymore ?

File: extlibs/vili/CMakeLists.txt (hunk 8 above)
Comment (Sygmei) [reply extlibs/vili/CMakeLists.txt:33]: Since `fmt` is required by both vili and ÖbEngineCore, I think the `find_package` should be located in the root `CMakeLists.txt` ?

**Review by Sygmei — APPROVED** [review]
Good job ! Thanks a lot for taking the time to integrate vcpkg into ÖbEngine :)

**Comment by TerensTare** [comment]
I noticed there were some conflicts between the `CMakeLists.txt` I submitted on `src/Core` and the one provided by the engine. Please suggest anything as this is the first time I resolve conflicts.
